iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+        }
+    }
+}
+
+/// State of a connected peer.
+#[derive(derive_more::Debug)]
+enum PeerState {
+    /// Peer is handling at least one request.
+    Busy {
+        #[debug("{}", active_requests.get())]
+        active_requests: NonZeroUsize,
+    },
+    /// Peer is idle.
+    Idle {
+        #[debug(skip)]
+        drop_key: delay_queue::Key,
+    },
+}
+
+/// Type of future that performs a download request.
+type DownloadFut = LocalBoxFuture<'static, (DownloadKind, Result<(), FailureAction>)>;
+
+#[derive(Debug)]
+struct Service<G: Getter, D: Dialer> {
+    /// The getter performs individual requests.
+    getter: G,
+    /// Map to query for peers that we believe have the data we are looking for.
+    providers: ProviderMap,
+    /// Dialer to get connections for required peers.
+    dialer: D,
+    /// Limits to concurrent tasks handled by the service.
+    concurrency_limits: ConcurrencyLimits,
+    /// Channel to receive messages from the service's handle.
+    msg_rx: mpsc::Receiver<Message>,
+    /// Peers available to use and their relevant information.
+    peers: HashMap<PublicKey, ConnectionInfo<D::Connection>>,
+    /// Queue to manage dropping keys.
+    goodbye_peer_queue: delay_queue::DelayQueue<PublicKey>,
+    /// Requests performed for download intents. Two download requests can produce the same
+    /// request. This map allows deduplication of efforts.
+    current_requests: HashMap<DownloadKind, ActiveRequestInfo>,
+    /// Downloads underway.
+    in_progress_downloads: FuturesUnordered<DownloadFut>,
+    /// Requests scheduled to be downloaded at a later time.
+    scheduled_requests: HashMap<DownloadKind, PendingRequestInfo>,
+    /// Queue of scheduled requests.
+    scheduled_request_queue: delay_queue::DelayQueue<DownloadKind>,
+}
+
+impl<G: Getter<Connection = D::Connection>, D: Dialer> Service<G, D> {
+    fn new(
+        getter: G,
+        dialer: D,
+        concurrency_limits: ConcurrencyLimits,
+        msg_rx: mpsc::Receiver<Message>,
+    ) -> Self {
+        Service {
+            getter,
+            providers: ProviderMap::default(),
+            dialer,
+            concurrency_limits,
+            msg_rx,
+            peers: HashMap::default(),
+            goodbye_peer_queue: delay_queue::DelayQueue::default(),
+            current_requests: HashMap::default(),
+            in_progress_downloads: FuturesUnordered::default(),
+            scheduled_requests: HashMap::default(),
+            scheduled_request_queue: delay_queue::DelayQueue::default(),
+        }
+    }
+
+    /// Main loop for the service.
+    async fn run(mut self) {
+        loop {
+            // check if we have capacity to dequeue another scheduled request
+            let at_capacity = self
+                .concurrency_limits
+                .at_requests_capacity(self.in_progress_downloads.len());
+
+            tokio::select! {
+                Some((peer, conn_result)) = self.dialer.next() => {
```

Comment:
shouldn't this have a fallback for the `select!` statement, given that you use `Some()` matching on the branches?

Reply:
Sorry, I meant in general for all the `Some(..)` matches in the `select!`, I usually have fallback clause to break out of the loop and log an error, if any of them actually return `None`, to make sure it shuts down if the streams are shut down

Reply:
ah I see, yes I'll add that, thanks

Reply:
I'm now more or less certain that this not necessary. 

Reading the docs of `tokio::select` when an enabled branch is ready, and the pattern doesn't match it's simply considered disabled and another branch is polled. Basically the only case in which we would land in a default branch is if all patters fail (which means, if all branches are disabled). Since there is a branch that does not do pattern matching (the one that polls the message channel), we should never reach that

lmk if this sounds convincing enough

Here are the docs I'm basing this, for completeness

> The select! macro accepts one or more branches with the following pattern:
> 
> `<pattern> = <async expression> (, if <precondition>)? => <handler>,`
>
> (...more docs here..)
> 
> The complete lifecycle of a select! expression is as follows:
>
>  1. Evaluate all provded `<precondition>` expressions. If the precondition returns false, disable the branch for the remainder of the current call to select!. Re-entering `select!` due to a loop clears the "disabled" state.
    2.  Aggregate the `<async expression>`s from each branch, including the disabled ones. If the branch is disabled, `<async expression>` is still evaluated, but the resulting future is not polled.
    3.  Concurrently await on the results for all remaining `<async expression>`s.
    4.  Once an `<async expression>` returns a value, attempt to apply the value to the provided `<pattern>`, if the pattern matches, evaluate `<handler>` and return. If the pattern does not match, disable the current branch and for the remainder of the current call to select!. Continue from step 3.
    5. If all branches are disabled, evaluate the else expression. If none is provided, panic.